_posts/2020_04_29_encrypted_keras_with_syft/encrypted_keras_with_syft.Rmd

```diff
@@ -0,0 +1,413 @@
+
+#### Secure multi-party computation
+
+In SMPC, we need a system of several agents who collaborate to provide a result no single agent could provide alone: "normal"
+computations (like addition, multiplication ...) on "secret" (encrypted) data. The assumption is that these agents are "honest
+but curious" -- honest, because they won't tamper with their share of data; curious in the sense that if they *were* (curious,
+that is), they wouldn't be able to inspect the data because it's encrypted.
+
+The principle behind this is *secret sharing*. A single piece of data -- a salary, say -- is "split up" into meaningless
+(hence, encrypted) parts which, when put together again, yield the original data. Here is an example.
+
+Say the parties involved are Julia, Greg, and me. The below function encrypts a single value, assigning to each of us their
+"meaningless" share:
+
+```{r}
+# a big prime number
+# all computations are performed in a finite field, for example, the integers modulo that prime
+Q <- 78090573363827
+ 
+encrypt <- function(x) {
+  # all but the very last share are random 
+  julias <- runif(1, min = -Q, max = Q)
+  gregs <- runif(1, min = -Q, max = Q)
+  mine <- (x - julias - gregs) %% Q
+  list (julias, gregs, mine)
+}
+
+# some top secret value no-one may get to see
+value <- 77777
+
+encrypted <- encrypt(value)
+encrypted
+```
+
+    [[1]]
+    [1] 7467283737857
+
+    [[2]]
+    [1] 36307804406429
+
+    [[3]]
+    [1] 34315485297318
+
+Once the three of us put our shares together, getting back the plain value is straightforward:
+
+```{r}
+decrypt <- function(shares) {
+  Reduce(sum, shares) %% Q  
+}
+
+decrypt(encrypted)
+```
+
+    77777
+
+As an example of how to compute on encrypted data, here's addition. (Other operations will be a lot less straightforward.) To
+add two numbers, just have everyone add their respective shares:
+
+```{r}
+add <- function(x, y) {
+  list(
+    # julia
+    (x[[1]] + y[[1]]) %% Q,
+    # greg
+    (x[[2]] + y[[2]]) %% Q,
+    # me
+    (x[[3]] + y[[3]]) %% Q
+  )
+}
+  
+x <- encrypt(11)
+y <- encrypt(122)
+
+decrypt(add(x, y))
+```
+
+    133
+
+Back to the setting of deep learning and the current task to be solved: Have the server apply gradient updates without ever
+seeing them. With secret sharing, it would work like this:
+
+Julia, Greg and me each want to train on our own private data. Together, we will be responsible for gradient averaging, that
+is, we'll form a *cluster* of *workers* united in that task. Now, the model owner *secret shares* the model, and we start
+training, each on their own data. After some number of iterations, we use secure averaging to combine our respective
+gradients. Then, all the server gets to see is the mean gradient, and there is no way to determine our respective
+contributions.
+
+#### Beyond private gradients
+
+Amazingly, it is even possible to *train* on encrypted data -- amongst others, using that same technique of secret sharing. Of
```

Comment:
Do you mean the same technique of secret sharing, among others. If yes, I'd reorder the sentence to reflect that. If no, I'm not sure what the "others" is referring to.